### PR TITLE
feat: add validation

### LINKED ISSUE #2: add message/resource validation
a validator is needed to validate the message/resource structure and the specific message/resource data while parsing

### PR DESCRIPTION
**Pull request by ethanwlee**
# Summary
closes https://github.com/TBD54566975/tbdex-dart/issues/2

- [x] validating messages/resources
  - introduces a validator using [json_schema](https://pub.dev/packages/json_schema) to validate the message/resources and the message/resource kinds in `parse()`
  - validator pulls in schemas from the `tbdex` submodule (inspired by [tbdex-js](https://github.com/TBD54566975/tbdex-js/pull/129))
- [x] verifying rfq offering requirements
  - completes `verifyOfferingRequirements()` in `rfq.dart` (except verifying the claims)

also adds ci

### CODE REVIEW

File: lib/src/protocol/validator.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  static Future<RefProvider> _createRefProvider(String schemasPath) async {
+    final schemaJson = json.decode(
+      await File(p.join(schemasPath, 'definitions.json')).readAsString(),
+    );
+
+    return RefProvider.sync(
+      (ref) => ref == 'https://tbdex.dev/definitions.json' ? schemaJson : null,
+    );
+  }
```
Comment (ethanwlee) [inline lib/src/protocol/validator.dart]: this is used to get the `definitions.json` $ref working as documented [in the json_schema spec](https://github.com/Workiva/json_schema?tab=readme-ov-file#synchronous-creation-local-ref-cache)

File: lib/src/protocol/models/message.dart [hunk 2]
```diff
@@ -75,15 +78,22 @@ abstract class Message {
   String? signature;
 
   static Future<Message> parse(String payload) async {
+    // TODO: how to initialize properly?
+    await Validator.initialize();
```
Comment (ethanwlee) [inline lib/src/protocol/models/message.dart]: would like a better solution for initializing `Validator`... any ideas?

Comment (fingersonfire) [reply lib/src/protocol/models/message.dart]: You beat me to it, I was just about to suggest possibly using a singleton with a static constructor. Nice work!

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+          cache: true
+
+      - name: Install Dependencies
+        run: just get
+
+      - name: Run Static Analysis
+        run: just analyze
+
+      - name: Run Tests
+        run: just test
```
Comment (wesbillman) [inline .github/workflows/ci.yml:30]: Nice!!

File: lib/src/protocol/models/message_data.dart [hunk 4]
```diff
@@ -184,13 +188,15 @@ class CloseData extends MessageData {
...
-      'reason': reason,
+      if (success != null) 'success': success,
+      if (reason != null) 'reason': reason,
     };
   }
 }
 
 class OrderData extends MessageData {
+  OrderData();
+
```
Comment (wesbillman) [inline lib/src/protocol/models/message_data.dart]: Do you need this here?

File: lib/src/protocol/models/resource.dart [hunk 5]
```diff
@@ -61,17 +64,28 @@ abstract class Resource {
...
+    }
+    Validator.validate(jsonResource, 'resource');
+
+    final jsonResourceData = jsonResource['data'];
+    final resourceMetadata = jsonResource['metadata'] as ResourceMetadata?;
+    if (resourceMetadata == null) {
+      throw Exception('metadata property is required');
+    }
+    final resourceKind = resourceMetadata.kind.name;
+    Validator.validate(jsonResourceData, resourceKind);
```
Comment (wesbillman) [inline lib/src/protocol/models/resource.dart]: This logic seems duplicated, would it be better to put it in a class that can be parsed with `data` and `metadata`?

File: lib/src/protocol/validator.dart [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  static final Map<String, JsonSchema> _schemaMap = {};
+  static final Validator _instance = Validator._internal();
+
+  static void validate(Map<String, dynamic> json, String schemaName) {
+    _instance.validateJson(json, schemaName);
+  }
+
+  Validator._();
```
Comment (wesbillman) [inline lib/src/protocol/validator.dart]: nit: but maybe we should move this up to just below the `_instance` definition to keep related things together?

**Review by wesbillman — APPROVED** [review]
Awesome!

File: lib/src/protocol/models/resource.dart (hunk 5 above)
Comment (ethanwlee) [reply lib/src/protocol/models/resource.dart]: will make sure to cover this in https://github.com/TBD54566975/tbdex-dart/issues/6!

**Pull request merged**